File: LongChallenge_2019_01_DPAIRS.py
# code_report Solution
# Video Link: https://youtu.be/N7FBSGN0W24
# Problem Link: https://www.codechef.com/JAN19A/problems/DPAIRS/

n, m = (int(i) for i in input().split())
a = [[int(x), i] for i, x in enumerate(input().split())]
b = [[int(x), i] for i, x in enumerate(input().split())]
a.sort()
b.sort()
seen = set()
front = False
i, j = 0, len(a)-1
while(len(seen) < n + m - 1):
    front = not front
    x = a[i] if front else a[j]
    for y in b:
        val = x[0] + y[0]
        if val not in seen:
            print(x[1], ' ', y[1])
            seen.add(val)
        if (len(seen) == n + m - 1):
            break;
    if front: i += 1
    else: j -= 1

File: LongChallenge_2019_01_DPAIRS.cpp
// code_report Solution
// Video Link: https://youtu.be/N7FBSGN0W24
// Problem Link: https://www.codechef.com/JAN19A/problems/DPAIRS/

#include <unordered_set>
#include <iostream>
#include <map>

using namespace std;

int main() {
   int n, m, x; cin >> n >> m;
   map<int, int> a, b;
   for (int i = 0; i < n; ++i) cin >> x, a[x] = i;
   for (int i = 0; i < m; ++i) cin >> x, b[x] = i;

   unordered_set<int> seen;
   auto front = false;
   auto i = begin(a), j = --end(a);
   while (seen.size() < n + m - 1) {
      front = !front;
      auto x = front ? i : j;
      for (auto y : b) {
         auto it = seen.insert(x->first + y.first);
         if (it.second) 
            cout << x->second << ' ' << y.second << endl;
         if (seen.size() == n + m - 1)
            break;
      }
      if (front) ++i; else --j;
   }

   return 0;
}
